ReceivedMessageListener class and DEMO

albertoursino opened the pull request:
Hi guys this is the last class of gruppo4's **basic** library.
This class can examine if a [smslibrary](https://github.com/Cogno-IDU/smslibrary.git)'s received message is a valid ring command.

DEMO apk [here](https://www.mediafire.com/file/794evw0gjduaas7/DEMO_2.0.apk/file) (Updated to 7a29327)

I changed the minSdkVersion to 23 in the commit n° 53eee4a because some "requiresAPI" in the permissions methods was making me nervous but nothing bad, we can change back to 21 and leave the reqAPI...

During testing the DEMO I discovered 2 bugs, fixed by me in the commits 135be10 and 7a29327.

Commit:
Created class file

Commit:
Created the class ReceivedMessageListener

Commit:
Created a DEMO to test our basic application libray

minSdkVersion from 21 to 23.

Commit:
Important FIX to the class RingCommandHandler

I discovered this bug by testing the application: the method "parseCommand" must add the SPLIT_CHARACTER to the SMSMessage data.

- Corrected the MainActivity accordingly
- Implemented the PasswordManager class

Commit:
Little specification fix

Commit:
Resolved the issue "https://github.com/Cogno-IDU/killer-app/issues/19"

SMSHandler wasn't receiveing the "ReceivedMessageListener" after the "onRequestPermissionsResult"

tonin-alessandra commented on app/src/main/java/com/eis/phoneringer/structure/ReceivedMessageListener.java:
```diff
@@ -0,0 +1,56 @@
+import com.eis.smslibrary.listeners.SMSReceivedListener;
+
+/**
+ * Listener used to manages incoming sms and evaluates if they are valid commands for the application
```
I think this should be "Listener used to **manage** incoming **SMS** and **evaluate** if..."

tonin-alessandra commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
+                Toast.makeText(context, "Command sent to " + phoneNumber.getText().toString(), Toast.LENGTH_SHORT).show();
+            }
+        };
+        //Passing to the AppManager a new RingCommand built through the phone Number the user given
```
I think "the phone number given by the user" should be better.

tonin-alessandra commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
+    }
+
+    /**
+     * Checks if the user grant the SMS permissions
```
Maybe the verb "to grant" should be used in the past tense?

tonin-alessandra commented:
I have nothing else to say, I think this is nice and simple code. Good job!

Commit:
Fixed some spelling errors

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
+        phoneNumber = findViewById(R.id.phone_number);
+        passwordManager = new PasswordManager(context);
+
+        /**
```
Double asterisks are used to indicate a part of the javadoc for this class, regular multiline comments should use only one asterisk.

albertoursino replied:
Can't I use specifications for other parts of the code other than methods or classes?

giovannivelludo replied:
Double asterisk are used to tell the javadoc tool to add that comment to the documentation, but your comment is just a comment and not the specification of a method or a description of a class, so it should use only one asterisk in the first line

albertoursino replied:
Oh you are right, I will change. Thanks.

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
 public class MainActivity extends AppCompatActivity {
 
+    private static Button ringButton = null;
+    private static EditText phoneNumber = null;
```
why are these `static`?

albertoursino replied:
Because it seems correct to me since it makes no sense to instantiate the same one twice, for example, EditText.
Am I wrong? Ready to learn.

CremaLuca replied:
I don't think those should be static either: the Activity is a normal instance and could have different views (those set on `setContentView(R.layout.*)`) therefore graphic elements shouldn't be static

giovannivelludo replied:
`ringButton` is only used in method `onCreate()`, so you could instantiate it there as a local variable, and `phoneNumber` could be left where it is, but I don't understand why you made it static. Does it need to be accessed and have the same value in different instances of MainActivity? Are we even going to have multiple instances of MainActivity needing to share that value? If not, then I would remove the `static` keyword. Android Studio always warns that placing Context classes in static fields is a memory leak, so it should be avoided unless absolutely necessary.

AlbertoUrsino replied:
I'll try your solution @giovannivelludo

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
+            SMSHandler.getInstance().setReceivedListener(new ReceivedMessageListener(getApplicationContext()));
+        } else {
+            finish();
+            System.exit(0);
```
what do these two lines do? do they terminate execution of the app? `finish()` terminates the current Activity, while I have no idea of what `System.exit(0)` does in Android, in Java it terminates execution of the current JVM.
why are you terminating the current Activity?

albertoursino replied:
Well I don't know sincerely.
Time ago I saw using this method of "exit" by @CremaLuca and since then I have been using it

CremaLuca replied:
My fault, I read the wrong answer on this https://stackoverflow.com/questions/6330200/how-to-quit-android-application-programmatically
Should have used `finishAndRemoveTask()` since we're using API 21+.

giovannivelludo replied:
What I would do in that `else` block would be display a Toast telling the user that we need those permissions in order for the app to work, and then call `requestPermission()` again. Closing the app seems a bit too drastic, and the user won't know why it happened.

albertoursino replied:
I'll try your solution @giovannivelludo

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
+        SMSSentListener smsSentListener = new SMSSentListener() {
+            @Override
+            public void onSMSSent(SMSMessage message, SMSMessage.SentState sentState) {
+                Toast.makeText(context, "Command sent to " + phoneNumber.getText().toString(), Toast.LENGTH_SHORT).show();
```
here I would use string resources from strings.xml, they should be used for text displayed to the user because they enable translations of the app, I'm not sure if we should worry about this in a demo though.

albertoursino replied:
Are u saying this for `"Command sent to "` ?

CremaLuca replied:
It'd be a great idea, we haven't used them before so we still have to get comfortable with those.

giovannivelludo replied:
> Are u saying this for `"Command sent to "` ?

Yes

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
+                !(getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED))
+            return false;
+        else
+            return true;
```
you could remove all of this and instead use
```
return (getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED) &&
(getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED)
```

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
 package com.eis.phoneringer;
 
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresApi;
```
you're not using this

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
+import android.Manifest;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.os.Build;
```
you're not using this

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/structure/ReceivedMessageListener.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    /**
+     * Creates the RingCommand object upon receipt of the message and calls the class {@link AppManager} in order to manage it only if it isn't null
```
"in order to manage" what? The RingCommand object or the message? I would specify what I'm referring to instead of using "it", repetitions are not a problem in documentation if they help make it more clear

albertoursino replied:
I'll specify better.

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/structure/ReceivedMessageListener.java:
```diff
@@ -0,0 +1,56 @@
+    private Context context;
+
+    /**
+     * Constructor which define a context
```
defines

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
+     *
+     * @param context The current application context
+     */
+    private void sendCommand(final Context context) {
```
why does this method require passing a Context? since it's located in an Activity it can use the Activity itself as a Context, or get the application context.

albertoursino replied:
You are in right but I didn't want to write everytime `getApplicationCode()`

giovannivelludo replied:
Assign the context once and use that variable

CremaLuca commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
+    private static Button ringButton = null;
+    private static EditText phoneNumber = null;
+    private static final String DEFAULT_PASSWORD = "password";
+    private PasswordManager passwordManager = null;
```
Why are you assigning `null` on a `null` variable? Those assignments look useless, is there a reason why you do this?

albertoursino replied:
For a better clarity of the code. 
Should I to remove them?

CremaLuca commented on app/src/main/java/com/eis/phoneringer/structure/RingCommandHandler.java:
```diff
@@ -49,12 +49,13 @@ public RingCommand parseMessage(SMSMessage smsMessage) {
      */
     public SMSMessage parseCommand(RingCommand ringCommand) {
-        return new SMSMessage(ringCommand.getPeer(), ringCommand.getPassword());
+        return new SMSMessage(ringCommand.getPeer(), SPLIT_CHARACTER + ringCommand.getPassword());
```
I still don't get why this is called SPLIT_CHARACTER if it doesn't split anything, you could name it something like ring command prefix or whatever: what if we decide to have more commands and use this class to parse them? Think big

giovannivelludo replied:
It splits different fields to allow parsing of the message, right?

CremaLuca replied:
But there are no _different fields_, there is just the password right now as field, the sms outcome is: "_password"

giovannivelludo replied:
lol, you're right

albertoursino replied:
Because in the parseMessage method the SMSMessage content is splitted thanks to the "_" character.
And think if we have to send a command with the password AND Idk, the volume we want to play the ringtone: we are going to use 2 SPLIT_CHARACTER like that: _password_100 where 100 is the max volume...

giovannivelludo replied:
I would modify parseMessage and make it use [`split(SPLIT_CHARACTER)`](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#split-java.lang.String-) to separate content by looking for `SPLIT_CHARACTER` inbetween different elements

d-u-d-e replied:
Why don't check for _ as the first character and then get the password as substring(1)? You are not splitting anything as @CremaLuca already noted.

albertoursino replied:
As you wish.
I will change the name from "SPLIT_CHARACTER" to "COMMAND_IDENTIFIER"

albertoursino replied:
@giovannivelludo the method already uses `split(SPLIT_CHARACTER)`

giovannivelludo replied:
sorry, I'm retarded. then you should change
```
return new RingCommand(smsMessage.getPeer(), parts[1]);
```
to
```
return new RingCommand(smsMessage.getPeer(), parts[0]);
```
if you're removing the `_` character from before the password (unless `parts[0]` is empty because of no separator, I can't check right now), or not use `split()` at all for now. With no `_` character the whole text of the SMSMessage is the password.

albertoursino replied:
And if we are using our library for another application?

albertoursino replied:
Messages can get confused

giovannivelludo replied:
> And if we are using our library for another application?

What library?

albertoursino replied:
[smslibrary](https://github.com/Cogno-IDU/smslibrary)

giovannivelludo replied:
I thought the purpose of `SPLIT_CHARACTER` was separating different fields in a message, not being an identifier of this application.  
If you want to use a custom code for this app you should use your own [`MessageParseStrategy`](https://github.com/Cogno-IDU/smslibrary/blob/master/smslibrary/src/main/java/com/eis/communication/MessageParseStrategy.java).  
My personal opinion is that this unnecessarily overcomplicates things, because it would be easier to do the parsing directly in the app like you did, without having to create an entire class that extends `MessageParseStrategy`, then calling `SMSMessageHandler.getInstance().setMessageParseStrategy()`, and then having to call `SMSMessageHandler`'s methods everytime you want to check if a message is meant for this app.
So I agree with using `SPLIT_CHARACTER` to identify messages meant for this app, but then I would change the name of that variable if separating fields is not its purpose.

albertoursino replied:
Take a look to this commit 918ead5

giovannivelludo replied:
Sorry, this conversation wasn't marked as resolved and the code on top of it was the old one

CremaLuca reviewed (changes requested):
A few fixes, looks good anyway

Commit:
Fixes after Velludo and Crema's reviews

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,126 @@
+ */
 public class MainActivity extends AppCompatActivity {
 
+    private Button ringButton;
```
this can be a local variable inside method onCreate(), it's used only there

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,126 @@
+     */
+    public boolean checkPermission() {
+        return !((getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED) ||
+                !(getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED));
```
This is too complicated to read and I think it might be wrong, what you want to do is check if both equivalences are true. What you're doing is ¬(P∨¬Q), but it should be P∧Q. Actually, you're doing the same equivalence twice. This should just be
```
return getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED;
```

d-u-d-e replied:
he probably wanted to type Manifest.permission.SEND_SMS. Anyway this is wrong

Commit:
Hotfixes

- Changed name of "SPLIT_CHARACTER" to "COMMAND_IDENTIFIER"
- Rewrote some specifications
- Fixed checkPermission method in the MainActivity
- Moved some variables
- Fixed a spelling error in strings.xml file

CremaLuca commented on app/src/main/java/com/eis/phoneringer/structure/RingCommandHandler.java:
```diff
@@ -32,29 +32,32 @@ public static RingCommandHandler getInstance() {
-            return new RingCommand(smsMessage.getPeer(), parts[1]);
+        final int PASSWORD_INDEX = 1;
+        String messageContent = smsMessage.getData();
+        if ((messageContent.charAt(0) + "").equals(COMMAND_IDENTIFIER)) {
```
Could probably have changed `COMMAND_IDENTIFIER` to be a `char`

CremaLuca commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -1,14 +1,127 @@
+                public void run() {
+                    requestPermission();
+                }
+            }, WAIT_TIME_PERMISSION);
```
To be 100% sure that the time is enough I'd have used `Toast.LENGTH_SHORT + WAIT_TIME_PERMISSION`, where extra time is short

albertoursino replied:
`Toast.LENGTH_SHORT` return an integer value equals to zero.
If I put it where u want, nothing will change, `WAIT_TIME_PERMISSION` is enough.

CremaLuca commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,8 @@
 <resources>
     <string name="app_name">PhoneRinger</string>
+    <string name="permissions_denied_toast">App needs permissions</string>
+    <string name="invalid_phone_number_toast">Invalid phone number</string>
```
SUPER NICE, string.xml rules

CremaLuca commented on app/src/main/java/com/eis/phoneringer/structure/RingCommandHandler.java:
```diff
@@ -32,29 +32,32 @@ public static RingCommandHandler getInstance() {
-            String[] parts = smsMessage.getData().split(SPLIT_CHARACTER);
-            //parts[0] is empty, parts[1] contains the password
-            return new RingCommand(smsMessage.getPeer(), parts[1]);
+        final int PASSWORD_INDEX = 1;
```
Could move this to be a class variable

CremaLuca reviewed (approved):
Nice, little modifications if you want to do them, not necessary

Commit:
Some changes requested by Luca Crema

Commit:
Changed method "parseMessage"

Now this method extracts the password using "substring" as Marco Mariotto suggested

AlbertoUrsino commented:
Now I think that this class can be merged

CremaLuca commented:
Three of the reviewers are from the same workgroup, I think at least another review from someone of another group is needed

giovannivelludo commented:
@CremaLuca I approved this pull request

Cogno-Marco commented on app/src/main/java/com/eis/phoneringer/structure/ReceivedMessageListener.java:
```diff
@@ -0,0 +1,57 @@
+     * Creates the {@link RingCommand} upon receipt of the message and calls the class {@link AppManager} if the RingCommand is not null
+     *
+     * @param smsMessage The SMSMessage object received
+     * @throws IllegalPasswordException Exception thrown by the method "onRingCommandReceived" of the class {@link AppManager} when the password received is not valid
+     */
```
In the specifications you say you throw `IllegalPasswordException`, but later on you just catch it and launch a Toast. You should either remove the throws from the specifications or remove the catch/toast and actually launch the exception

Cogno-Marco commented on app/src/main/java/com/eis/phoneringer/structure/ReceivedMessageListener.java:
```diff
@@ -0,0 +1,57 @@
+        //Let's parse this message and build a RingCommand object
+        RingCommand ringCommand = RingCommandHandler.getInstance().parseMessage(smsMessage);
+
+        if (ringCommand == null) {
+            Log.d("Invalid RingCommand", "The message received is not a ring command");
```
I would throw an exception here too since you log that this is an error.

albertoursino replied:
Yes I agree with you.

But... How can we catch this exception in another class for example in the MainActivity?

Cogno-Marco replied:
oh you're right, didn't think of that.
I guess we'll keep it like this 👍

Cogno-Marco reviewed (changes requested):
A couple of points in a function. The rest is alright 👍

Commit:
Removed useless parentheses

Cogno-Marco reviewed (approved):
Everything has been fixed 👍
